Add Ergonomic event macros, closes #39 and #74.

davidbarsky opened the pull request:
Partially addresses #39. I can created macros for other levels, but I personally think the primary levels that we should support are trace/info and error.

hawkw commented:
@davidbarsky 
> I personally think the primary levels that we should support are trace/info and error.

I disagree. I think it would be nice to have an event macro for each of the log macros that the log crate supports.

Ideally, it would be possible for invoking an event macro with no fields to use the same syntax as invoking the log macro, so that you could begin porting a library to `tokio-trace` by simply switching the dependency; however, i'm not sure if this is really possible due to how the format arguments are captured. Regardless, providing a direct translation of all the `log` crate's macros will make the experience for newcomers easier.

I know we've debated this in the past, but personally, I think there's value for having multiple levels of verbosity --- I think having separate "trace" and "debug" log levels is a good thing, as well as having separate "error" and "warning" levels. Libraries like tokio emit so many log records at the trace level that having a separate debug level for the less-verbose messages is extremely useful, and I suspect this is the case in a tracing context as well.

davidbarsky commented:
> I know we've debated this in the past, but personally, I think there's value for having multiple levels of verbosity --- I think having separate "trace" and "debug" log levels is a good thing, as well as having separate "error" and "warning" levels. Libraries like tokio emit so many log records at the trace level that having a separate debug level for the less-verbose messages is extremely useful, and I suspect this is the case in a tracing context as well.

That's very fair. My concern it that while we're making it easier for users of traditional loggers to adopt, I haven't really seen any tracing libraries introduce this concept of "levels". That being said, I'll update this PR to introduce the other levels.

hawkw commented:
@davidbarsky BTW, this is looking good; I've only held off on merging it because I'm in the process of preparing a branch that splits out the `tokio-trace-core` crate from `tokio-trace`, and I think we're going to want to rebase this on top of that.

davidbarsky commented:
no worries. that sounds good!

On Wed, Oct 31, 2018 at 6:31 PM Eliza Weisman <notifications@github.com>
wrote:

> @davidbarsky <https://github.com/davidbarsky> BTW, this is looking good;
> I've only held off on merging it because I'm in the process of preparing a
> branch that splits out the tokio-trace-core crate from tokio-trace, and I
> think we're going to want to rebase this on top of that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/hawkw/tokio-trace-prototype/pull/62#issuecomment-434870069>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB-NPurDfttuFihVPJhg7y2k5aTroilGks5uqiSxgaJpZM4YC43l>
> .
>

davidbarsky commented on tokio-trace-tower/src/lib.rs:
```diff
@@ -27,8 +27,8 @@ where
     type Response = T::Response;
-    type Future = Instrumented<T::Future>;
     type Error = T::Error;
+    type Future = Instrumented<T::Future>;
```
> This change seems unrelated, can we back it out?

Right, sorry.

> I'm assuming this is due to us running differing versions of rustfmt --- are you on nightly?

 No, this was IntelliJ's lint saying that the implementation was out of order with the defintion. Do you want me to revert?

hawkw replied:
> Do you want me to revert?

As long as it still passes `rustfmt`, I don't _really_ care. In general I try to keep unrelated formatting changes out of commits that add or modify code so that the git history is clearer, but this is so minor it doesn't really matter to me.

davidbarsky commented on tokio-trace/src/lib.rs:
```diff
@@ -116,6 +116,26 @@ macro_rules! event {
+}
+
+#[macro_export]
+macro_rules! error {
```
Yep, just need a land a few deliverables at work first.

hawkw replied:
Alright, take your time!

davidbarsky replied:
Fixed

davidbarsky commented:
Finally got around to fixing feedback. Changes:

- Rebased this branch atop of #67.
- Added `trace`, `warn`, and `debug` macros.
- Addressed #74. It's possible now to have `event!`/`log`-compatible macros in that don't require passing an empty `{}`.

hawkw commented:
@davidbarsky I think we need to rebase this branch onto `master` rather than #67?

davidbarsky commented:
> @davidbarsky I think we need to rebase this branch onto master rather than #67?
Hmm, I thought I did rebase atop of master? In fairness, I'm also terrible at using Git.

> Before merging, can we update the commit message to say that this closes #39 and #74? Thanks!
Can do! I can add this as part of the squash message?

davidbarsky commented:
>> Rebased this branch atop of #67.
> Before merging, can we update the commit message to say that this closes #39 and #74? Thanks!

Sorry! I misspoke. This was rebased atop of master.
